[Kernel] Update the usage docs to reflect the recent API changes (#2444)
delta-io/delta: kernel/USER_GUIDE.md

  FilteredColumnarBatch scanFileColumnarBatch = fileIter.next();

>   CloseableIterator<Row> scanFileRows = scanFileColumnarBatch.getRows();

Reviewer: Remove this line? We define `scanFileRows` twice -- here and in the try statement

  // Get the physical read schema of columns to read from the Parquet data files
  StructType physicalReadSchema =
    ScanStateRow.getPhysicalDataReadSchema(tableClient, scanState);

  try (CloseableIterator<Row> scanFileRows = scanFileColumnarBatch.getRows()) {
    while (scanFileRows.hasNext()) {
      Row scanFileRow = scanFileRows.next();

      // From the scan file row, extract the file path, size and modification time metadata
      // needed to read the file.

>       FileStatus fileStatus = InternalScanFileUtils.getAddFileStatus(scanFileRow);

Reviewer: In a general sense... if we are including this API in an example it seems like it shouldn't be in an internal util package?

      // Open the scan file which is a Parquet file using connector's own
      // Parquet reader or default Parquet reader provided by the Kernel (which
      // is used in this example).
      CloseableIterator<ColumnarBatch> physicalDataIter =
        tableClient.getParquetHandler().readParquetFiles(
          singletonCloseableIterator(fileStatus),
          physicalReadSchema,
          Optional.empty() /* optional predicate the connector can apply to filter data from the reader */
        );

      // Now the physical data read from the Parquet data file is converted to a table
      // logical data. Logical data may include the addition of partition columns and/or
      // subset of rows deleted
      try (
         CloseableIterator<FilteredColumnarBatch> transformedData =
           Scan.transformPhysicalData(
             tableClient,
             scanState,
             scanFileRow,
             physicalDataIter)) {
        while (transformedData.hasNext()) {

>           FilteredColumnarBatch filteredData = transformedData.next();

Reviewer: I think the name `filteredData` isn't helpful. I feel like it should reflect that it's the transformed batch, or something to do with it being the logical data

            // check if the row is selected or not
            if (selectionVector.isEmpty() || selectionVector.get().getBoolean(rowIndex)) {
              // Assuming the column type is String.
              // If it is a different type, call the relevant function on the `ColumnVector`
              System.out.println(column0.getString(rowIndex));
            }
          }

	  // access the data for column at ordinal 1
	  ColumnVector column1 = dataBatch.getColumnVector(1);
	  for (int rowIndex = 0; rowIndex < column1.getSize(); rowIndex++) {
            // check if the row is selected or not
            if (selectionVector.isEmpty() || selectionVector.get().getBoolean(rowIndex)) {
              // Assuming the column type is Long.
              // If it is a different type, call the relevant function on the `ColumnVector`
              System.out.println(column1.getLong(rowIndex));
            }
          }
	  // .. more ..
        }
      }
    }
  }
}
```

A few working examples to read Delta tables within a single process are available [here](https://github.com/delta-io/delta/tree/master/kernel/examples).

#### Important Note

> * All the Delta protocol-level details are encoded in the rows returned by `Scan.getScanFiles` API, but you do not have to understand them in order to read the table data correctly. All you need is to get the physical Parquet file details from the scan file row, read the data from the Parquet file, and return in `ColumnarBatch` format. The physical data can be converted into the data in the logical data of the table using [`Scan.transformPhysicalData`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Scan.html#transformPhysicalData-io.delta.kernel.client.TableClient-io.delta.kernel.data.Row-io.delta.kernel.data.Row-io.delta.kernel.utils.CloseableIterator-). Transformation to logical data involves adding partition column data, deleting rows that are marked as deleted by the scan file deletion vector. In the future, there will be more information (i.e. more columns) added to these logical data returned, but your code will not have to change to accommodate those changes. This is the major advantage of the abstractions provided by Delta Kernel.

Reviewer: maybe "... get the physical Parquet file status from the scan file row" to keep our wording consistent?

> * All the Delta protocol-level details are encoded in the rows returned by `Scan.getScanFiles` API, but you do not have to understand them in order to read the table data correctly. All you need is to get the physical Parquet file details from the scan file row, read the data from the Parquet file, and return in `ColumnarBatch` format. The physical data can be converted into the data in the logical data of the table using [`Scan.transformPhysicalData`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Scan.html#transformPhysicalData-io.delta.kernel.client.TableClient-io.delta.kernel.data.Row-io.delta.kernel.data.Row-io.delta.kernel.utils.CloseableIterator-). Transformation to logical data involves adding partition column data, deleting rows that are marked as deleted by the scan file deletion vector. In the future, there will be more information (i.e. more columns) added to these logical data returned, but your code will not have to change to accommodate those changes. This is the major advantage of the abstractions provided by Delta Kernel.

Reviewer: ```suggestion
* All the Delta protocol-level details are encoded in the rows returned by `Scan.getScanFiles` API, but you do not have to understand them in order to read the table data correctly. All you need is to get the Parquet file status from each scan file row and read the data from the Parquet file into the `ColumnarBatch` format. The physical data is converted into the logical data of the table using [`Scan.transformPhysicalData`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Scan.html#transformPhysicalData-io.delta.kernel.client.TableClient-io.delta.kernel.data.Row-io.delta.kernel.data.Row-io.delta.kernel.utils.CloseableIterator-). Transformation to logical data involves adding partition column data and deleting rows that are marked as deleted by the scan file deletion vector. In the future, there will be more information (i.e. more columns) added to these logical data returned, but your code will not have to change to accommodate those changes. This is the major advantage of the abstractions provided by Delta Kernel.
```

Reviewer: > Transformation to logical data involves adding partition column data and deleting rows that are marked as deleted by the scan file deletion vector. In the future, there will be more information (i.e. more columns) added to these logical data returned, but your code will not have to change to accommodate those changes. This is the major advantage of the abstractions provided by Delta Kernel.

Do we want to be more vague about what this transformation does now + in the future what it may do? I feel like the important take-away is that the transformation takes care of any protocol-level operations and future protocol changes will be included in this transformation.

Author: I think providing some examples of transformations should be ok, it tells the connector developer what as sort of example transformations are done.

Reviewer: Can we make it more reflective that those are examples of transformations then? We don't have to update it again in the future

i.e. "Transformation to logical data includes steps like adding partition column data, ... "

* Credentials and permissions: The connector must populate its `FileSystemClient` with the necessary configurations and credentials for the client to retrieve the necessary data from the storage system. For example, an implementation based on Hadoop's FileSystem abstractions can be passed S3 credentials via the Hadoop configurations.
* Decryption: If file system objects are encrypted, then the implementation must decrypt the data before returning the data.

#### Step 2.3: Implement [`ParquetHandler`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ParquetHandler.html)

As the name suggests, this interface contains everything related to reading Parquet files. It has been designed such that a connector can plug in a wide variety of implementations, from a simple single-threaded reader to a very advanced multi-threaded reader with pre-fetching and advanced connector-specific expression pushdown. Let's explore the methods to implement, and the guarantees associated with them. 

> ##### Method [`readParquetFiles()`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ParquetHandler.html#readParquetFiles-io.delta.kernel.utils.CloseableIterator-io.delta.kernel.types.StructType-)

Reviewer: include the method arguments in this title maybe? the below paragraph says "The columns to be read from the Parquet file are defined by the physical schema" but to a first time reader what is this physical schema and where is it provided/defined?

This method takes as input `FileStatus`s which contains metadata such as file path, size etc. of the Parquet file to read. The columns to be read from the Parquet file are defined by the physical schema. To implement this method, you may have to first implement your own [`ColumnarBatch`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/ColumnarBatch.html) and [`ColumnVector`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/ColumnVector.html) which is used to represent the in-memory data generated from the Parquet files.

When identifying the columns to read, note that there are multiple types of columns in the physical schema (represented as a [`StructType`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/types/StructType.html)). 

> * Data columns: Columns that are expected to be read from the Parquet file. Based on the `StructField` object defining the column, read the column in the Parquet file that matches the same name or field id. If the column has a field id (as `parquet.field.id` in column field metadata) then it is used to match the column in Parquet file otherwise name is used for matching.

Reviewer: ```suggestion
* Data columns: Columns that are expected to be read from the Parquet file. Based on the `StructField` object defining the column, read the column in the Parquet file that matches the same name or field id. If the column has a field id (stored as `parquet.field.id` in the `StructField` metadata) then the field id should be used to match the column in the Parquet file. Otherwise, the column name should be used for matching.
```

This method takes as input `FileStatus`s of the JSON files and returns the data in a series of columnar batches. The columns to be read from the JSON file are defined by the physical schema, and the return batches must match that schema. To implement this method, you may have to first implement your own [`ColumnarBatch`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/ColumnarBatch.html)and [`ColumnVector`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/ColumnVector.html) which is used to represent the in-memory data generated from the JSON files.

When identifying the columns to read, note that there are multiple types of columns in the physical schema (represented as a [`StructType`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/types/StructType.html)). 
 
##### Method [`parseJson`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/JsonHandler.html#parseJson-io.delta.kernel.data.ColumnVector-io.delta.kernel.types.StructType-)

> This method allows parsing a `ColumnVector` of string values which are in JSON format into the output format specified by the `outputSchema`. If a given column column in `outputSchema` is not found, then a null value is returned. It optionally takes a selection vector which indicates what entries in the input `ColumnVector` of strings to parse. If an entry is not selected then a `null` value is returned as parsed output for that particular entry in the output.

Reviewer: ```suggestion
This method allows parsing a `ColumnVector` of string values which are in JSON format into the output format specified by the `outputSchema`. If a given column in `outputSchema` is not found, then a null value is returned. It optionally takes a selection vector which indicates what entries in the input `ColumnVector` of strings to parse. If an entry is not selected then a `null` value is returned as parsed output for that particular entry in the output.
```

Reviewer: Same thing as before about adding the method arguments to the headers. Otherwise what is `outputSchema` here?

* When applicable (for example, with Java Kernel APIs), you have to make sure to call the close() method as you consume the `ColumnarBatch`es of scan files (that is, either serialize the rows or use them to read the table data).

#### Step 3.3: Distribute the file information to the workers

> If you are building a connector for a distributed engine like Spark/Presto/Trino/Flink, then your connector has to send all the scan metadata from the query planning machine (henceforth called the driver) to task execution machines (henceforth called the workers). You will have to serialize and deserialize the scan state and scan file rows. It is the connector job to implement serialization and deserialization utilities for a [`Row`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/Row.html). If the connector wants to split reading one scan files into multiple tasks, it can add additional connector specific split context to the task. At the task, connector can use its own Parquet reader to read the specific part of the file indicated by the split info.

Reviewer: ```suggestion
If you are building a connector for a distributed engine like Spark/Presto/Trino/Flink, then your connector has to send all the scan metadata from the query planning machine (henceforth called the driver) to task execution machines (henceforth called the workers). You will have to serialize and deserialize the scan state and scan file rows. It is the connector job to implement serialization and deserialization utilities for a [`Row`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/Row.html). If the connector wants to split reading one scan file into multiple tasks, it can add additional connector specific split context to the task. At the task, the connector can use its own Parquet reader to read the specific part of the file indicated by the split info.
```

Byte[] scanStateRowBytes = RowUtils.serialize(scanStateRow);
Byte[] scanFileRowBytes = RowUtils.serialize(scanFileRow);

> // Optionally the connector add a split info to the task (scan file, scan state) to

Reviewer: ```suggestion
// Optionally the connector adds a split info to the task (scan file, scan state) to
```

> // split reading of the Parquet files into multiple tasks. The task gets split info

Reviewer: ```suggestion
// split reading of a Parquet file into multiple tasks. The task gets split info
```

// Additional option predicate such as dynamic filters the connector wants to
// pass to the reader when reading files.
Predicate optPredicate = ...;

// Get the physical read schema of columns to read from the Parquet data files
StructType physicalReadSchema =
  ScanStateRow.getPhysicalDataReadSchema(tableClient, scanStateRow);

// From the scan file row, extract the file path, size and modification metadata
// needed to read the file.
FileStatus fileStatus = InternalScanFileUtils.getAddFileStatus(scanFileRow);

// Open the scan file which is a Parquet file using connector's own
// Parquet reader which supports reading specific parts (split) of the file.
// If the connector doesn't have its own Parquet reader, it can use the
// default Parquet reader provider which at the moment doesn't support reading
// a specific part of the file, but reads the entire file from the beginning.
CloseableIterator<ColumnarBatch> physicalDataIter =
  connectParquetReader.readParquetFile(
  fileStatus
  physicalReadSchema,
  split, // what part of the Parquet file to read data from
  optPredicate /* additional predicate the connector can apply to filter data from the reader */
);

> // Now the physical data read from the Parquet data file is converted to a table

Reviewer: "a table logical data"?

Author: updated to: "to logical data the table represents"
